Toggle switch component guidelines (#249)
primer/design: content/components/toggle-switch.mdx

    <img
        width="456"
        role="presentation"
        src="https://user-images.githubusercontent.com/2313998/159587116-aa07a847-a633-4515-82b3-20ebe2349914.png"
    />
</Box>

### Progressive disclosure

<Box mb={3} display="flex" alignItems="flex-start" flexDirection={["column", "column", "column", "column", "row"]} sx={{gap: 3}}>
    <Box as="p" mt="0">The <a href="/ui-patterns/progressive-disclosure">progressive disclosure pattern</a> may be used to hide or show content based on whether a toggle switch is on.</Box>
    <img
        width="456"
        role="presentation"
        src="https://user-images.githubusercontent.com/2313998/159591632-f604ec1d-aca9-42f7-ad65-16f8811f192f.png"
    />
</Box>

## Accessibility

### Keyboard navigation

Toggle switches should have the same keyboard navigation features as a button. A toggle switch may be focused using the `Tab` key, and pressing the `Enter` or `Space` key will toggle the switch.

### Touch target

> When the UI is rendered on a touch-screen device, ensure that the toggle switch has a large enough touch target size. The toggle switch should react when the user taps or hovers anywhere in the touch target area, even if it isn't directly on the control.

Reviewer: "Toggle switch should react when the user taps or hovers anywhere in the touch target area". I'm thinking you mean the cursor should switch to a clickable cursor but it almost sounds like it will switch on hover. Maybe we can clarify?"

My suggestion doesn't sound quite right just yet but something similar? 

```suggestion
When the UI is rendered on a touch-screen device, ensure that the toggle switch has a large enough touch target size. The toggle switch should be able to be activated when the user taps or hovers anywhere in the touch target area, even if it isn't directly on the control.
```

---
title: Toggle switch
description: A toggle switch is used to immediately toggle a setting on or off.
---

import {Box, Text} from '@primer/components'

## Anatomy

> 

Reviewer 2: I would change the anatomy image to be full width, even if it's a small component, and add the list of components below it.

<Box display="flex" alignItems="flex-start" flexDirection={["column", "column", "column", "column", "row"]} sx={{gap: 3}}>
    <Box as="ul" m="0" p="0">
        <li><strong>Track</strong>: The clickable area where the toggle switch moves back and forth</li>
        <li><strong>Knob</strong>: Changes color and position depending on whether the switch is on or off</li>
        <li><strong>Status</strong>: A text label showing whether the switch is on or off</li>
    </Box>
    <img
        width="456"
        role="presentation"
        src="https://user-images.githubusercontent.com/2313998/159587114-97d7de3c-9ff6-4f13-beca-6e4da2d548c6.png"
    />
</Box>

> ## Contextual content

Reviewer 2: This is all about labels, could the heading just be "Labels"?

I would also suggest that this section becomes part of the Options section, and that you move the Options section to here, so it's:
- Anatomy
- Options
  - Labels
  - Layout
  - Size
  - Loading
  - Progressive disclosure
- Mixed values
- Toggle switch vs checkbox
- Accessibility
  - Keyboard navigation
  - Touch target

Potentially **Accessibility** could be before **Mixed values** too.

<Box mb={5} display="flex" alignItems="flex-start" flexDirection={["column", "column", "column", "column", "row"]} sx={{gap: 3}}>
    <div>
    <Box as="p" mt="0">Every toggle switch should have a label that says what is being turned on or off. Optionally, you may include a description about the setting that is being turned on or off.</Box>

>     <Box as="p" mt="0">Labels should be succinct, and can describe a object or an action (for example: "Discussions" or "Automatically watch repositories"). Don't use phrases that describe the switch's state in your label.</Box>

Reviewer 2: ```suggestion
    <Box as="p" mt="0">Labels should be succinct, and can describe an object or an action (for example: "Discussions" or "Automatically watch repositories"). Don't use phrases that describe the switch's state in your label.</Box>
```

    </div>
    <img
        width="456"
        role="presentation"
        src="https://user-images.githubusercontent.com/2313998/159587115-5ffcdefb-f5e9-4180-8dc3-5328b0440f7f.png"
    />
</Box>

<DoDontContainer>
  <Do>
    <img role="presentation" src="https://user-images.githubusercontent.com/2313998/159587088-99e8fe4c-59a3-4452-836d-bd6f2297045a.png" />
    <Caption>Use succinct phrases with verbs or nouns</Caption>
  </Do>
  <Dont>
    <img role="presentation" src="https://user-images.githubusercontent.com/2313998/159587096-5f8a6913-df95-4ea0-9838-a366faf95e52.png" />
    <Caption>Don't use phrases that describe the switch's state</Caption>
  </Dont>
</DoDontContainer>

> ## Representing mixed values

Reviewer 2: "Mixed values" to simplify?

Since a toggle switch can only be "on" or "off", it could be confusing to use a switch to represent mixed values.

> If you have a section of configuration options that can be disabled, use a button instead of a toggle switch. Disabling the section turns off all configuration options, but remembers their previous state when the user wants to turn the section back on. While the section is disabled, the configuration options are hidden.

Reviewer 2: ```suggestion
If you have a group of configuration options that can be disabled, use a button instead of a toggle switch. Disabling the group turns off all configuration options, but remembers their previous state when the user wants to turn that section back on. While the section is disabled, the configuration options are hidden.
```

> ![Configuration options section that can be enabled or disabled. Configuration controls are hidden when it is disabled.](https://user-images.githubusercontent.com/2313998/159587120-2d842830-18c7-4429-bd10-f5424f6fac9e.png)

Reviewer 2: Maybe move image above the previous paragraph, "If you have a..."

<DoDontContainer>
  <Do>
    <img role="presentation" src="https://user-images.githubusercontent.com/2313998/159587100-1a1af5b2-dd19-456a-a0ed-c2f9fa7f7c2e.png" />
    <Caption>Use a button to affect multiple related toggles</Caption>
  </Do>
  <Dont>
    <img role="presentation" src="https://user-images.githubusercontent.com/2313998/159587101-446548c2-41c3-45c1-b710-db3149e2ebe4.png" />
    <Caption>Don't use a toggle switch to represent mixed values</Caption>
  </Dont>
</DoDontContainer>

## Toggle switch vs checkbox

Toggle switches are buttons that change and save a boolean configuration option when they are toggled.

> Checkboxes are form inputs that are used to select one or more items from a list, and are not saved without explicit confirmation such as pressing a "save" button.

Reviewer 2: ```suggestion
Checkboxes are form inputs that are used to select one or more items from a list, and are not saved without explicit confirmation such as pressing a **Save** button.
```

A toggle switch should never be used as a replacement for a checkbox. See the [saving patterns](/ui-patterns/saving) for more details.

<DoDontContainer>
  <Do>
    <img role="presentation" src="https://user-images.githubusercontent.com/2313998/159587105-45d50974-eb76-46cd-a3fd-c1e0f605f3ef.png" />
    <Caption>Use a toggle switch like a button</Caption>
  </Do>
  <Dont>
    <img role="presentation" src="https://user-images.githubusercontent.com/2313998/159587113-6c8461f7-3d4d-4e6a-b3a5-c585c5143fba.png" />
    <Caption>Don't use a toggle switch as a replacement for a checkbox</Caption>
  </Dont>
</DoDontContainer>

## Options

### Layout

<Box mb={3} display="flex" alignItems="flex-start" flexDirection={["column", "column", "column", "column", "row"]} sx={{gap: 3}}>
    <div>

>         <Box as="p" mt="0">By default, lay out a toggle switch horizontally justified with it's label and optional description.</Box>

Reviewer 2: ```suggestion
        <Box as="p" mt="0">By default, lay out a toggle switch horizontally justified with its label and optional description.</Box>
```

    />
</Box>

### Loading

<Box mb={3} display="flex" alignItems="flex-start" flexDirection={["column", "column", "column", "column", "row"]} sx={{gap: 3}}>
    <Box as="p" mt="0">When a toggle switch is waiting for something to load before toggling, a loading indicator should be displayed. The switch should not prematurely toggle from "on" to "off". Wait until loading is complete before showing a change.</Box>
    <img
        width="456"
        role="presentation"
        src="https://user-images.githubusercontent.com/2313998/159587116-aa07a847-a633-4515-82b3-20ebe2349914.png"
    />
</Box>

### Progressive disclosure

<Box mb={3} display="flex" alignItems="flex-start" flexDirection={["column", "column", "column", "column", "row"]} sx={{gap: 3}}>
    <Box as="p" mt="0">The <a href="/ui-patterns/progressive-disclosure">progressive disclosure pattern</a> may be used to hide or show content based on whether a toggle switch is on.</Box>
    <img
        width="456"
        role="presentation"
        src="https://user-images.githubusercontent.com/2313998/159591632-f604ec1d-aca9-42f7-ad65-16f8811f192f.png"
    />
</Box>

## Accessibility

### Keyboard navigation

> Toggle switches should have the same keyboard navigation features as a button. A toggle switch may be focused using the `Tab` key, and pressing the `Enter` or `Space` key will press the button.

Reviewer 2: Press the _button_ or press the _knob_?

Author: I was referring to "button" as in the `<button>` element, but I see how that's not clear.

I think it would be clearer to say "will toggle the switch".

### Touch target

> When the UI is rendered on a touch-screen device, ensure that the toggle switch has a sufficient touch target size. The toggle switch should react when the user taps or hovers anywhere in the touch target area, even if it isn't directly on the control.

Reviewer 2: ```suggestion
When the UI is rendered on a touch-screen device, ensure that the toggle switch has a large enough touch target size. The toggle switch should react when the user taps or hovers anywhere in the touch target area, even if it isn't directly on the control.
```

I don't love my suggestion, but "sufficient" sounds weird.
